Support reusable connections

juanjoDiaz opened the pull request:
Second attempt to add reusable connections for AMQP.

This breaks the API in a minimal way. The fix is to wrap the `AmqpConnectionSettings` with a `DefaultAmqpConnectionProvider`.
Or even use an implicit conversion from `AmqpConnectionSettings => DefaultAmqpConnectionProvider(AmqpConnectionSettings)`.

The get/release methods should be thread-safe now. Not sure if that's good enough or you have something else in .

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -0,0 +1,105 @@
+/**
+ * Only for internal implementations
+ */
+sealed trait AmqpConnectionProvider {
```
`@DoNotInherit`

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -0,0 +1,105 @@
+}
+
+final case class DefaultAmqpConnectionProvider(settings: AmqpConnectionSettings) extends AmqpConnectionProvider {
+  override def get = settings.getConnection
```
Hmmm, shouldn't the connection logic go here rather than live in the settings now that we have a `provider`?

juanjoDiaz replied:
It could but then each provider will need to match against all possible settings.
It's more reliable if each settings can provide its own connection so all providers work with all settings.

Other option is what I did in my original PR. Get rid of settings and consider them as providers and the reusable provider just a provider that wrap other providers.

johanandren replied:
Hmm, maybe it's more of a naming thing, but, now we have a provider that does not provide anything but has a special case where it caches connections, and a `Settings` that creates/provides connections, this makes me think we are missing something with the modeling ;)

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -0,0 +1,105 @@
+
+final case class ReusableAmqpConnectionProvider(settings: AmqpConnectionSettings, automaticRelease: Boolean = true)
+    extends AmqpConnectionProvider {
+  var clients = 0
```
I think we discussed this before, this is not threadsafe. Make an internal state ADT like so:

```
private sealed trait State
private case object Empty extends State
privat case object Connecting extends State
private final case class Connected(connection: Connection, clients: Int)
private case object Closing extends State
private val state = new AtomicReference[State](Empty)
```

And do CAS on it for all updates like so: 
```
@tailrec
override def get: Connection = {
  state.get match {
     case Empty =>
        if (state.compareAndSet(Empty, Connecting)) {
          // only one thread will get here, so now we can safely connect
          ... cas swap Connected in ...
       } else get // another thread is connecting, try again
    case Connecting => get // another thread is connecting, try again
    case c @ Connected(connection, clients) => 
       if (state.compareAndSet(c, Connected(connection, clients + 1)) connection
       else get // try again
    ... and so on...
```

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -0,0 +1,105 @@
+    ReusableAmqpConnectionProvider(settings: AmqpConnectionSettings, automaticRelease)
+}
+
+final case class ReusableAmqpConnectionProvider(settings: AmqpConnectionSettings, automaticRelease: Boolean = true)
```
Add scaladoc for what `automaticRelease` does

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionSettings.scala:
```diff
@@ -0,0 +1,166 @@
+ */
+case class AmqpConnectionLocal() extends AmqpConnectionSettings {
+  override def getConnection: Connection = new ConnectionFactory().newConnection
+}
```
No need for a class, use a single `object` extending `AmqpConnectionSettings` instead.

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionSettings.scala:
```diff
@@ -0,0 +1,166 @@
+ * Only for internal implementations
+ */
+sealed trait AmqpConnectionSettings {
+  private[amqp] def getConnection: Connection
```
Shouldn't this be `createConnection` or `newConnection` rather, it should never do any caching, right?

juanjoDiaz replied:
I can rename it. It doesn't do caching.

johanandren commented on amqp/src/test/scala/akka/stream/alpakka/amqp/scaladsl/AmqpConnectionProvidersSpec.scala:
```diff
@@ -0,0 +1,129 @@
+      val connection2 = connectionProvider.get
+      connection1 should not equal connection2
+      connectionProvider.release(connection1)
+      connectionProvider.release(connection2)
```
As these calls to get and release can happen concurrently, I think it would be valuable if the tests were concurrent as well. Will not be deterministic but would hopefully have caught the concurrency bug I noted above.

johanandren commented on docs/src/main/paradox/amqp.md:
```diff
@@ -12,6 +12,19 @@ The AMQP connector provides Akka Stream sources and sinks to connect to AMQP ser
+
+There are two types of @scaladoc[AmqpConnectionProvider](akka.stream.alpakka.amqp.AmqpConnectionProvider):
+* @scaladoc[DefaultAmqpConnectionProvider](akka.stream.alpakka.amqp.DefaultAmqpConnectionProvider) which create a new connection for each connector and releases it when the stage stops.
+* @scaladoc[ReusableAmqpConnectionProvider](akka.stream.alpakka.amqp.ReusableAmqpConnectionProvider) which reuses the same connections for all stages. By default it closes the connection whenever the last stage using the provider stops. Optionally, it takes `automaticRelease` boolean parameter so the connection is not automatically release and the user have to release it explicitly.
```
How could the user release it explicitly when `release` checks the boolean flag?

juanjoDiaz replied:
Release never check a boolean flag any more since that was specific for the cached connection and we needed a generic release method.
The reusable connection offers a client counter so you can do `provider.clients.forEach(_ => provider.release)` or something like that

juanjoDiaz commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -0,0 +1,105 @@
+
+  override def get =
+    cachedConnectionRef
+      .updateAndGet(
```
Doesn't this atomic way of updating guarantee that only a thread is running at once?
Similarly to your proposed solution.

johanandren replied:
The problem is that you side effect inside of the function (see the java doc of `updateAndGet` explicitly saying you cannot do that)

johanandren replied:
The clients update could interleave or even not be seen ever between other threads, meaning that the number is pretty much arbitrary. In addition to that the connect call could be called multiple times in case of contention leading to connections being created and leaked.

juanjoDiaz commented:
I've completely removed the connection settings and left only connection providers. As you noted, it made no sense having both. I've also renamed the providers to a more consistent naming schema and updated the docs.

I've implemented the non-blocking logic as you suggested.

The only thing left from your requests is to make the test concurrent.
Are you suggesting that I wrap all the creation/release of connections in Futures so they are run in parallel?
Some tests `release -> check -> release -> check`. So if they are concurrent I can't make the checks. Any idea is welcome.

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -6,100 +6,210 @@ package akka.stream.alpakka.amqp
+    case Empty =>
+      if (state.compareAndSet(Empty, Connecting)) {
+        val connection = provider.get
+        state.compareAndSet(Connecting, Connected(connection, 0))
```
Client count should be 1 here

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -6,100 +6,210 @@ package akka.stream.alpakka.amqp
+  @tailrec
+  override def release(connection: Connection): Unit = state.get match {
+    case c @ Connected(cachedConnection, clients) =>
+      if (state.compareAndSet(c, Closing)) {
```
Check the cached connection as the first thing in `release` instead of modifying the state and then checking.

Don't change to closing unless the clients are 0 and automatic release is true, for other cases it should just subtract one from the client count.

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -0,0 +1,216 @@
+        throw new IllegalArgumentException("Can't release a connection that's not owned by this provider")
+
+      if (state.compareAndSet(c, Closing)) {
+        if (clients == 0 || !automaticRelease) {
```
What thread will see it get 0 if we are the last client and only subtract in the other branch? It should also not do the Closing swap unless it wants to close, which should only be if clients - 1 becomes 0 and automaticRelease, this you can check before doing the cas.

juanjoDiaz replied:
I've updated this to be 1 instead of 0. So when there is only 1 client left, the connection is closed.

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -0,0 +1,216 @@
+          provider.release(connection)
+          state.compareAndSet(Closing, Empty)
+        } else {
+          state.compareAndSet(Closing, Connected(cachedConnection, clients - 1))
```
So what happens if we just lost this cas because another thread also got here?

juanjoDiaz replied:
How could that happen?
Doesn't `state.compareAndSet(c, Closing)` in line 206 ensures that only 1 thread can go into this code?

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -0,0 +1,216 @@
+          state.compareAndSet(Closing, Connected(cachedConnection, clients - 1))
+        }
+      } else release(connection)
+    case _ => release(connection)
```
I think it would be good to not match-all-recurse here but match explicitly, and for the unexpected cases throw an IllegalStateException, so that we get to know if there is a case here that we didn't think of.

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -0,0 +1,216 @@
+      if (cachedConnection != connection)
+        throw new IllegalArgumentException("Can't release a connection that's not owned by this provider")
+
+      if (state.compareAndSet(c, Closing)) {
```
It should not always switch to `Closing`, only if it is the last client and automatic release is enabled it should try to win the cas for `Closing`, in the other case it should just update state.compareAndSet(c, c.copy(clients = c.clients - 1))` and it must check if that succeeded and recurse if it failed (because another client may have `released` or called `get` concurrently)

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -191,26 +191,31 @@ final case class AmqpCachedConnectionProvider(provider: AmqpConnectionProvider,
-      } else release(connection)
-    case _ => release(connection)
+      } else {
+        if (!state.compareAndSet(Closing, Connected(cachedConnection, clients - 1))) release(connection)
```
The cas should not be from `Closing` here.  We observed `c` and we want to update the value based on the assumption that the state didn't change, if we see that assumption was wrong (cas failed) we recurse.

This is what makes closing work if there are two concurrent clients releasing concurrently, both sees `clients == 2` and comes down here, tries to cas, one wins, while the other recurses and sees `clients == 1` going for the close-branch.

juanjoDiaz replied:
true!

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -191,26 +191,31 @@ final case class AmqpCachedConnectionProvider(provider: AmqpConnectionProvider,
+      if (clients == 1 || !automaticRelease) {
+        if (state.compareAndSet(c, Closing)) {
           provider.release(connection)
           state.compareAndSet(Closing, Empty)
```
This cas should not fail, but if it did we'd have a bug, so please check the result and throw illegal state if it failed here.

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -191,26 +191,31 @@ final case class AmqpCachedConnectionProvider(provider: AmqpConnectionProvider,
+        if (!state.compareAndSet(Closing, Connected(cachedConnection, clients - 1))) release(connection)
+      }
+    case Closing => release(connection)
+    case _ => throw new RuntimeException("Unknown provider state.")
```
Throw `IllegalState` and include the _unexpected_ state.

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -191,26 +191,31 @@ final case class AmqpCachedConnectionProvider(provider: AmqpConnectionProvider,
           state.compareAndSet(Closing, Empty)
-        } else {
-          state.compareAndSet(Closing, Connected(cachedConnection, clients - 1))
         }
```
If we lost this cas to get to close, we need to recurse

juanjoDiaz replied:
You mean `state.compareAndSet(Closing, Empty)`?

Didn't you say above that you rather throwing an exception since we should never lose it?

johanandren replied:
I think GitHub made something weird with the offset, because now I'm not sure which one it relates to. I'll take another look at the complete source.

johanandren replied:
In general for *every* cas you do, you should look at the return value and act in some way. I think you are correct here, if it was changed from `Closing` to something else by another thread that is a bug, so throw IllegalState for that.

johanandren commented on amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala:
```diff
@@ -191,26 +191,31 @@ final case class AmqpCachedConnectionProvider(provider: AmqpConnectionProvider,
 
   @tailrec
   override def release(connection: Connection): Unit = state.get match {
+    case Empty => ()
```
Isn't this an illegal-state one as well, how could it be `Empty` when we release unless something is wrong?

juanjoDiaz replied:
If the user call release before the connection is even opened.

We could:
* Throw an exception stating that you can't release if the connection hasn't been opened.
* Just let it be. It's already released so there is nothing to do

I chose option B. Would you prefer option A?

johanandren replied:
Yes, absolutely option A, because it would either be a bug in our concurrency logic and be _us_ calling it or somehow a bug in the user logic, either way we'd not want it to silently behave as everything is ok but to detect that something was not what we thought it is.

juanjoDiaz commented:
Should be all good now.
